mllib/src/main/scala/org/apache/spark/mllib/stat/test/ADTest.scala

```diff
@@ -0,0 +1,264 @@
+    val n = data.count()
+    val makeDist = initDist(distName, data, n, params.toArray)
+    val localData = data.sortBy(x => x).mapPartitions(calcPartAD(_, makeDist, n)).collect()
+    val s = localData.foldLeft((0.0, 0.0)) { case ((prevStat, prevCt), (rawStat, adj, ct)) =>
+      val adjVal = 2 * prevCt * adj
+      val adjustedStat = rawStat + adjVal
+      val cumCt = prevCt + ct
+      (prevStat + adjustedStat, cumCt)
+    }._1
+    val ADStat = -1 * n - s / n
+    val criticalVals = makeDist().getCVs(n)
+    new ADTestResult(ADStat, criticalVals, NullHypothesis.oneSample.toString)
+  }
+
+
+  /**
+   * Calculate a partition's contribution to the Anderson Darling statistic.
+   * In each partition we calculate 2 values, an unadjusted value that is contributed to the AD
+   * statistic directly, a value that must be adjusted by the number of values in the prior
+   * partition, and a count of the elements in that partition
+   * @param part `Iterator[Double]` a partition of the data sample to be analyzed
+   * @param makeDist `() => ADTheoreticalDist` a function to create a class that extends the
+   *                ADTheoreticalDist trait, which requires various methods, used in creating 1
+   *                object per partition
+   * @param n `Double` the total size of the data sample
+   * @return `Iterator[(Double, Double, Double)]` The first element corresponds to the
+   *        position-independent contribution to the AD statistic, the second is the value that must
+   *        be scaled by the number of elements in prior partitions and the third is the number of
+   *        elements in this partition
+   */
+  def calcPartAD(part: Iterator[Double], makeDist: () => ADTheoreticalDist, n: Double)
+    : Iterator[(Double, Double, Double)] = {
+    val dist = makeDist()
+    val initAcc = (0.0, 0.0, 0.0)
+    val pResult = part.zipWithIndex.foldLeft(initAcc) { case ((prevS, prevC, prevCt), (v, i)) =>
+      val y = dist.cdf(v)
+      val a = math.log(y)
+      val b = math.log(1 - y)
+      val unAdjusted = a * (2 * i + 1) + b * (2 * n - 2 * i - 1)
+      val adjConstant = a - b
+      (prevS + unAdjusted, prevC + adjConstant, prevCt + 1)
+    }
+    Array(pResult).iterator
+  }
+
+  /**
+   * Create a function to produce a distribution per partition.
+   * The user must provide parameters to initialize the distribution (note that the critical values
+   * assume that the parameters were estimated from the data).
+   * @param distName `String` name of distribution
+   * @param data `RDD[Double]` sample of data to analyze (and hence to use for parameter estimation
+   *            where possible)
+   * @param n `Double` size of data sample
+   * @param params `Double*` Initialization parameters for distribution
+   * @return `() => RealDistribution` function to create distribution object
+   */
+  def initDist(distName: String, data: RDD[Double], n: Double, params: Array[Double])
```

Comment:
Thanks, missed that when we changed it. Used to be there to estimate parameters for some of the distributions. I've actually reworked some of this code, since there is no issue now with passing the ADTheoreticalDist directly, as the math3 distributions I was using are serializable in 3.4.1. I'm cleaning up before pushing the latest version.